### PR TITLE
feat(core/xref): support data-lt attributes

### PR DESCRIPTION
**Pull request by sidvishnoi**
As a part of #1662.

Have a play with these inputs:
``` html
<!-- basic usage -->
<a data-lt="event handler">event handlers</a>
```

``` html
<!-- a bit advanced -->
<dfn data-lt="url parser">url parsing</dfn>
<a>URL parsing</a>
<a>URL parser</a>
<a>URL parsers</a>  <!-- this one should fail -->
```

Try following separately with `respecConfig.pluralize = true`
``` html
<dfn data-lt="url parser">url parsing</dfn>
<a>URL parsing</a>
<a>URL parser</a>
<a>URL parsers</a>  <!-- << still fail as we do not pluralize data-lt content -->
```

``` html
<dfn>URL parser</dfn>
<a>URL parsing</a>  <!-- should fail -->
<a>URL parser</a>
<a>URL parsers</a>  <!-- this one should now pass -->
```

``` html
<dfn data-lt="url parsing">URL parser</dfn>
 <!-- all should fail as "url parsing" is not defined externally. -->
<!-- if it was intended to be a local dfn,
   its closest ancestor should have empty data-cite -->
<a>URL parsing</a>
<a>URL parser</a>
<a>URL parsers</a>
```

``` html
<!-- if data-lt is present, first one is used for xref -->
<dfn data-lt="url parser|url parsing">URL parser</dfn> 
<!-- all should now pass -->
<a>URL parsing</a>
<a>URL parser</a>
<a>URL parsers</a> 
```

### CODE REVIEW

File: src/core/link-to-dfn.js [hunk 1]
```diff
@@ -80,6 +81,16 @@ export async function run(conf, doc, cb) {
         const dfn = titles[target.title][target.for];
         if (dfn[0].dataset.cite) {
           $ant[0].dataset.cite = dfn[0].dataset.cite;
+        } else if (
+          conf.xref &&
+          dfn[0].dataset.lt &&
+          dfn[0].dataset.lt.includes($ant[0].textContent.toLowerCase())
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: Here: 
```JS
dfn[0].dataset.lt.includes($ant[0].textContent.toLowerCase()
```

We might want:

```
dfn[0].dataset.lt.split("|").includes($ant[0].textContent.toLowerCase()
```

File: src/core/link-to-dfn.js [hunk 2]
```diff
@@ -80,6 +81,16 @@ export async function run(conf, doc, cb) {
...
+        ) {
+          const lt = dfn[0].dataset.lt.split("|");
+          const txt = $ant[0].textContent.toLowerCase();
+          const key = dfn[0].textContent.toLowerCase();
+          $ant[0].classList.add("xref");
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: we do make use of `data-dfn-type` elsewhere, maybe we can classify it as an "xref"?

File: src/core/link-to-dfn.js [hunk 3]
```diff
@@ -80,6 +81,16 @@ export async function run(conf, doc, cb) {
...
+          $ant[0].dataset.xref = lt[0] === pluralOf(txt) ? key : lt[0];
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: As we discussed on Slack... let's fix pluralize

File: src/core/link-to-dfn.js (hunk 2 above)
Comment (sidvishnoi) [reply src/core/link-to-dfn.js]: For searching, `.xref` would be faster I believe. But we can make use of `data-dfn-type` in other ways.

File: src/core/link-to-dfn.js [hunk 4]
```diff
@@ -73,15 +73,25 @@ export async function run(conf, doc, cb) {
...
+          : [];
+        const txt = ant.textContent.toLowerCase();
+        if (dfn.dataset.cite) {
+          ant.dataset.cite = dfn.dataset.cite;
+        } else if (conf.xref && (lt.includes(txt) || plurals.includes(txt))) {
+          ant.dataset.dfnType = "xref";
+          ant.dataset.xref = lt[0];
         } else {
-          const frag = "#" + encodeURIComponent(dfn.prop("id"));
+          const frag = "#" + encodeURIComponent($dfn.prop("id"));
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: Nit: `"#" + encodeURIComponent(dfn.id)`

File: src/core/link-to-dfn.js [hunk 5]
```diff
@@ -95,15 +105,15 @@ export async function run(conf, doc, cb) {
         // definitions that have either other text, or other
         // whitespace, inside the <dfn>.
         if (
-          dfn.closest("code,pre").length ||
-          (dfn.contents().length === 1 && dfn.children("code").length === 1)
+          $dfn.closest("code,pre").length ||
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: nit: can probably unjquery these now

File: src/core/link-to-dfn.js [hunk 6]
```diff
@@ -73,15 +73,25 @@ export async function run(conf, doc, cb) {
...
+        const lt = dfn.dataset.lt ? dfn.dataset.lt.split("|") : [];
+        const plurals = dfn.dataset.plurals
+          ? dfn.dataset.plurals.split("|")
+          : [];
+        const txt = ant.textContent.toLowerCase();
+        if (dfn.dataset.cite) {
+          ant.dataset.cite = dfn.dataset.cite;
+        } else if (conf.xref && (lt.includes(txt) || plurals.includes(txt))) {
+          ant.dataset.dfnType = "xref";
+          ant.dataset.xref = lt[0];
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: As discussed, a comment here explaining why first wins would be helpful.

File: src/core/link-to-dfn.js [hunk 7]
```diff
@@ -92,12 +92,13 @@ export async function run(conf, doc, cb) {
           // data-lt[0] serves as unique id for the dfn which this element references
           ant.dataset.xref = lt[0];
         } else {
-          const frag = "#" + encodeURIComponent($dfn.prop("id"));
-          $ant.attr("href", frag).addClass("internalDFN");
+          const frag = "#" + encodeURIComponent(dfn.id);
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: I think we should just do:

ant.href = "#" + encodeURIComponent(dfn.id);

Comment (marcoscaceres) [reply src/core/link-to-dfn.js]: update the above.

File: src/core/link-to-dfn.js [hunk 8]
```diff
@@ -92,12 +92,13 @@ export async function run(conf, doc, cb) {
...
+          ant.setAttribute("href", frag);
+          ant.classList.add("internalDFN");
         }
         // add a bikeshed style indication of the type of link
-        if (!$ant.attr("data-link-type")) {
-          $ant.attr("data-link-type", "dfn");
+        if (!ant.hasAttribute("data-link-type")) {
+          ant.setAttribute("data-link-type", "dfn");
```
Comment (marcoscaceres) [inline src/core/link-to-dfn.js]: same here...
```
ant.dataset.linkType = "dfn";
```

File: src/core/link-to-dfn.js (hunk 7 above)
Comment (sidvishnoi) [reply src/core/link-to-dfn.js]: no need to `encodeURIComponent`?

Comment (marcoscaceres) [reply src/core/link-to-dfn.js]: when we set it via the .href, it automatically happens.

File: src/core/link-to-dfn.js [hunk 9]
```diff
@@ -73,37 +73,50 @@ export async function run(conf, doc, cb) {
...
         // Note that contents().length===1 excludes
         // definitions that have either other text, or other
         // whitespace, inside the <dfn>.
+        // TODO: un-jquery-fy
         if (
-          dfn.closest("code,pre").length ||
-          (dfn.contents().length === 1 && dfn.children("code").length === 1)
+          dfn.closest("code,pre") ||
+          ($dfn.contents().length === 1 &&
+            [...dfn.children].filter(c => c.localName === "code").length === 1)
```
Comment (sidvishnoi) [inline src/core/link-to-dfn.js]: needs consideration

Comment (marcoscaceres) [reply src/core/link-to-dfn.js]: this is fine. We can update it when Edge get's scope support.

Comment (sidvishnoi) [reply src/core/link-to-dfn.js]: merging then :)

**Pull request merged**